File: topics/openacc/solutions/diffusion/diffusion2d_openacc_mpi.cpp
#include <iostream>
#include <cstdlib>
#include <mpi.h>
#include <openacc.h>

#include "diffusion2d.hpp"
#include "util.h"

// 2D diffusion example
// the grid has a fixed width of nx=128
// the use specifies the height, ny, as a power of two
// note that nx and ny have 2 added to them to account for halos

int main(int argc, char** argv) {
    // set up parameters
    // first argument is the y dimension = 2^arg
    size_t pow    = read_arg(argc, argv, 1, 8);
    // second argument is the number of time steps
    size_t nsteps = read_arg(argc, argv, 2, 100);
    // third argument is nonzero if shared memory version is to be used
    bool use_shared = read_arg(argc, argv, 3, 0);

    // set domain size
    size_t nx = 128;
    size_t ny = 1 << pow;
    double dt = 0.1;

    // initialize MPI
    int mpi_rank, mpi_size;

    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &mpi_rank);
    MPI_Comm_size(MPI_COMM_WORLD, &mpi_size);
    if (ny % mpi_size) {
        std::cout << "error : global domain dimension " << ny
                  << "must be divisible by number of MPI ranks "
                  << mpi_size << "\n";
        exit(1);
    } else if (mpi_rank == 0) {
        std::cout << "\n## " << mpi_size << " MPI ranks" << std::endl;
        std::cout << "## " << nx << "x" << ny
                  << " : " << nx << "x" << ny/mpi_size << " per rank"
                  << " for " << nsteps << " time steps"
                  << " (" << nx*ny << " grid points)\n";
    }

    ny /= mpi_size;

    // adjust dimensions for halo
    nx += 2;
    ny += 2;

    // allocate memory on device and host
    // note : allocate enough memory for the halo around the boundary
    auto buffer_size = nx*ny;

#ifdef OPENACC_DATA
    double *x0     = malloc_host_pinned<double>(buffer_size);
    double *x1     = malloc_host_pinned<double>(buffer_size);
#else
    double *x_host = (double *) malloc(buffer_size*sizeof(double));
    // double *x_host = malloc_host_pinned<double>(buffer_size);
    double *x0     = malloc_device<double>(buffer_size);
    double *x1     = malloc_device<double>(buffer_size);
#endif

    double start_diffusion, time_diffusion;

#ifdef OPENACC_DATA
    #pragma acc data create(x0[0:buffer_size]) copyout(x1[0:buffer_size])
#endif
    {
        // set initial conditions of 0 everywhere
        fill_gpu(x0, 0., buffer_size);
        fill_gpu(x1, 0., buffer_size);

        // set boundary conditions of 1 on south border
        if (mpi_rank == 0) {
            fill_gpu(x0, 1., nx);
            fill_gpu(x1, 1., nx);
        }

        if (mpi_rank == mpi_size-1) {
            fill_gpu(x0+nx*(ny-1), 1., nx);
            fill_gpu(x1+nx*(ny-1), 1., nx);
        }

        auto south = mpi_rank - 1;
        auto north = mpi_rank + 1;

        // time stepping loop
        #pragma acc wait
        start_diffusion = get_time();
        for(auto step=0; step<nsteps; ++step) {
            MPI_Request requests[4];
            MPI_Status  statuses[4];
            auto num_requests = 0;

#ifdef OPENACC_DATA
            #pragma acc host_data use_device(x0, x1)
#endif
            {
                if (south >= 0) {
                    // x0(:, 0) <- south
                    MPI_Irecv(x0,    nx, MPI_DOUBLE, south, 0, MPI_COMM_WORLD,
                              &requests[0]);
                    // x0(:, 1) -> south
                    MPI_Isend(x0+nx, nx, MPI_DOUBLE, south, 0, MPI_COMM_WORLD,
                              &requests[1]);
                    num_requests += 2;
                }

                // exchange with north
                if(north < mpi_size) {
                    // x0(:, ny-1) <- north
                    MPI_Irecv(x0+(ny-1)*nx, nx, MPI_DOUBLE, north, 0,
                              MPI_COMM_WORLD, &requests[num_requests]);
                    // x0(:, ny-2) -> north
                    MPI_Isend(x0+(ny-2)*nx, nx, MPI_DOUBLE, north, 0,
                              MPI_COMM_WORLD, &requests[num_requests+1]);
                    num_requests += 2;
                }
            }

            MPI_Waitall(num_requests, requests, statuses);

            diffusion_gpu(x0, x1, nx-2, ny-2, dt);

#ifdef OPENACC_DATA
            copy_gpu(x0, x1, buffer_size);
#else
            std::swap(x0, x1);
#endif
        }

        #pragma acc wait
        time_diffusion = get_time() - start_diffusion;
    } // end of acc data

#ifdef OPENACC_DATA
    auto x_res = x1;
#else
    copy_to_host<double>(x0, x_host, buffer_size);
    auto x_res = x_host;
#endif

    if (mpi_rank == 0) {
        std::cout << "## " << time_diffusion << "s, "
                  << nsteps*(nx-2)*(ny-2)*mpi_size / time_diffusion
                  << " points/second\n\n";

        std::cout << "writing to output.bin/bov\n";
        write_to_file(nx, ny, x_res);
    }

    MPI_Finalize();
    return 0;
}

File: topics/openacc/solutions/diffusion/diffusion2d_openacc_mpi.f90
program main
  use mpi
  use util
  use diffusion2d
  implicit none

  integer :: pow, nx, ny, nsteps, err, buffer_size, i, mpi_rank, mpi_size, &
       south, north, num_requests
  real(kind(0d0)) :: dt, start_diffusion, time_diffusion
  real(kind(0d0)), dimension(:), allocatable :: x0, x1
  integer, dimension(4) :: requests
  integer, dimension(4) :: statuses(MPI_STATUS_SIZE)

  call mpi_init(err)
  call mpi_comm_rank(MPI_COMM_WORLD, mpi_rank, err)
  call mpi_comm_size(MPI_COMM_WORLD, mpi_size, err)

  ! read arguments
  pow = read_arg(1, 8)
  nsteps = read_arg(2, 100)

  ! set domain size
  nx = 128
  ny = 2**pow
  dt = 0.1

  if (mod(ny, mpi_size) /= 0) then
     write(*, '(a i0 a i0)') 'error : global domain dimension ', ny, &
          'must be divisible by number of MPI ranks ', mpi_size;
     stop
  else if (mpi_rank == 0) then
     write(*, *) ''
     write(*, '(a i0 a)') '##', mpi_size, ' MPI ranks'
     write(*, '(a i0 a i0 a i0 a i0 a i0 a i0 a i0)') &
          '## ', nx, 'x', ny, ' : ', nx, 'x', ny/mpi_size, ' per rank for ', &
          nsteps, ' time steps (', nx*ny, ' grid points)'
  endif

  ny = ny / mpi_size
  nx = nx + 2
  ny = ny + 2
  buffer_size = nx*ny

  allocate(x0(buffer_size), x1(buffer_size), stat=err)
  if (err /= 0) then
     stop 'failed to allocate arrays'
  endif

  !$acc data create(x0) copyout(x1)

  ! set initial conditions of 0 everywhere
  call fill_gpu(x0, 0d0, buffer_size)
  call fill_gpu(x1, 0d0, buffer_size)

  ! set boundary conditions of 1 on south border
  if (mpi_rank == 0) then
     call fill_gpu(x0, 1d0, nx);
     call fill_gpu(x1, 1d0, nx);
  endif

  if (mpi_rank == mpi_size-1) then
     call fill_gpu(x0(nx*(ny-1)+1:nx*ny), 1d0, nx);
     call fill_gpu(x1(nx*(ny-1)+1:nx*ny), 1d0, nx);
  endif


  south = mpi_rank - 1
  north = mpi_rank + 1

  requests = 0
  statuses = 0

  !$acc wait
  start_diffusion = get_time()
  do i = 1, nsteps
     num_requests = 0
     !$acc host_data use_device(x0, x1)
     if (south >= 0) then
        call mpi_irecv(x0(1:), nx, MPI_DOUBLE, south, 0, MPI_COMM_WORLD, &
             requests(1), err)
        call mpi_isend(x0(nx+1:), nx, MPI_DOUBLE, south, 0, MPI_COMM_WORLD, &
             requests(2), err)
        num_requests = num_requests + 2
     endif

     if (north < mpi_size) then
        call mpi_irecv(x0((ny-1)*nx+1:), nx, MPI_DOUBLE, north, 0, &
             MPI_COMM_WORLD, requests(num_requests+1), err)
        call mpi_isend(x0((ny-2)*nx+1:), nx, MPI_DOUBLE, north, 0, &
             MPI_COMM_WORLD, requests(num_requests+2), err)
        num_requests = num_requests + 2
     endif
     !$acc end host_data

     call mpi_waitall(num_requests, requests, statuses, err)

     call diffusion_gpu(x0, x1, nx-2, ny-2, dt)

     call copy_gpu(x0, x1, buffer_size)
  enddo

  !$acc wait
  time_diffusion = get_time() - start_diffusion

  !$acc end data

  if (mpi_rank == 0) then
     write(*, '(a f0.6 a e0.6e2 a)') '## ', time_diffusion, 's, ', &
          real(nsteps)*(nx-2)*(ny-2) / time_diffusion, ' points/second'
     write(*, *) ''
     write(*, *) ''
     write(*, '(a)') 'writing to output.bin/bov'
     write(*, *) ''
  endif

  call write_to_file(nx, ny, x1);
  call MPI_Finalize(err)

end program main
